### DISCUSSION TITLE
Automatically set initial window size to fit contents

### DISCUSSION THREAD

**Discussion by archer884** [emilk/egui#2858] [Q&A]
Here's what I got:

<img width="344" alt="Screenshot 2023-03-30 at 7 05 18 PM" src="https://user-images.githubusercontent.com/679494/228990873-42eb57af-c500-4a6b-b8b6-a91e5aee31fd.png">

It's not bad, and I like that it wasn't that hard to create, but there's a problem: that extra space at the bottom is driving me crazy, and it changes size on different screens and operating systems and so forth. In fact, on some systems, this overflows. I would love to figure out some way to set the window's initial size based on what's IN the window rather than just setting it by trial and error in a run/edit loop, you know? But I haven't figured out a working way to do that.

This is less of a problem horizontally because I can trust the width of these elements to scale to a degree, but I'll admit that I have a similar problem with X and Y both.

Edit: I just realized that, technically, this isn't a window. It's built around `CentralPanel` instead. I just tried *using* `Window`, but that exhibits the same problem except that now there's a perfectly-sized `Window` inside the (not-so-perfectly-sized) application window, which is what I was referring to before, colloquially, as a "window."

Edit 2: It's really not essential that this be *automatic.* It would be fine to be able to know how much room the ui elements take up and then perform a resize by hand. I tried that at one point, actually, but the results weren't great. The size reported on the first frame is wrong; the size reported on the second frame appears to be correct, but by that time it seems to be too late to resize the window?

Edit the Third: I have now split the UI into two panels. With a footer panel at the bottom, it's definitely less obvious that there is "extra" space in the UI, and I have an element at the bottom of the central panel that can grow a little without looking too funny. This is a relatively satisfying result, albeit not completely, but I'd still like to know if there's some way to add up what the appropriate size for a given UI is. Also, yeah, if there's a smarter way to do this stuff, let me know!

**Comment by liuzhch1** [comment]
hi, have you found a better way to achieve this? I am also troubled by this problem. I searched for a long time and only found this [helpful discussion](https://github.com/emilk/egui/discussions/4329).

We can use this tricky way to make the window fit its content when it starts
- First time render with `egui::Window` which auto shrink to content size
- Get the size of the window
- Send `ViewportCommand::InnerSize` command to adjust the native window size
- Then render with `egui::CentralPanel` to get a native window

Here is the [code permalink](https://github.com/liuzhch1/eframe-window-size-fit-content/blob/cea02ceefe1bf04eeee9614d1c84761aee631398/src/main.rs#L1-L61)
```rs
fn main() {
  let app = AppGUI::default();
  eframe::run_native(
    "eframe window size fit content",
    eframe::NativeOptions::default(),
    Box::new(|_cc| Ok(Box::new(app))),
  )
  .unwrap();
}


struct AppGUI {
  name: String,
  age: u8,
  content: String,
  _first_render: bool,
}


impl Default for AppGUI {
  fn default() -> Self {
    Self {
      name: "John".to_string(),
      age: 20,
      content: "This demo shows how the eframe window size fit content works.".to_string(),
      _first_render: true,
    }
  }
}
impl AppGUI {
  fn pre_render(&mut self, ctx: &eframe::egui::Context) {
    egui::Window::new("pre_render")
      .title_bar(false)
      .fixed_pos((0.0, 0.0))
      .show(ctx, |ui| {
        self.render(ui);
      });
  }
  fn render(&mut self, ui: &mut egui::Ui) {
    ui.add(
      egui::Label::new(egui::RichText::new("Hello, world!").heading())
        .wrap_mode(egui::TextWrapMode::Extend),
    );
    ui.label(format!("Name: {}", self.name));
    ui.label(format!("Age: {}", self.age));
    ui.label(format!("{} ", self.content));
  }
}


impl eframe::App for AppGUI {
  fn update(&mut self, ctx: &eframe::egui::Context, _frame: &mut eframe::Frame) {
    if self._first_render {
      self.pre_render(ctx);
      self._first_render = false;
      let window_size = ctx.used_size();
      ctx.send_viewport_cmd(egui::ViewportCommand::InnerSize(window_size));
    } else {
      egui::CentralPanel::default().show(ctx, |ui| {
        self.render(ui);
      });
    }
  }
}
```
Not quite elegant, hope it helps you.

**Comment by PJaros** [comment]
@liuzhch1 This doesn't work for me. I've tried this with `eframe = { version = "0.33.0", features = ["default"] }` .
I'm running this under Linux with Gnome 49.1.

My current code:
```rust
use eframe::egui;

fn main() -> eframe::Result {
    let app = AppGUI::default();
    eframe::run_native(
        "eframe window size fit content",
        eframe::NativeOptions::default(),
        Box::new(|_cc| Ok(Box::new(app))),
    )
}

struct AppGUI {
    _first_render: bool,
}

impl Default for AppGUI {
    fn default() -> Self {
        Self {
            _first_render: true,
        }
    }
}

impl AppGUI {
    fn pre_render(&mut self, ctx: &eframe::egui::Context) {
        egui::Window::new("pre_render")
            .title_bar(false)
            .fixed_pos((0.0, 0.0))
            .show(ctx, |ui| {
                self.render(ui);
            });
    }
    fn render(&mut self, ui: &mut egui::Ui) {
        ui.add(
            egui::Label::new(egui::RichText::new("Hello, world!").heading())
                .wrap_mode(egui::TextWrapMode::Extend),
        );
        for n in 1..=40 {
            ui.label(format!("Line: {}", n));
        }
    }
}

impl eframe::App for AppGUI {
    fn update(&mut self, ctx: &eframe::egui::Context, _frame: &mut eframe::Frame) {
        if self._first_render {
            self.pre_render(ctx);
            self._first_render = false;
            let window_size = ctx.used_size();
            ctx.send_viewport_cmd(egui::ViewportCommand::InnerSize(window_size));
        } else {
            egui::CentralPanel::default().show(ctx, |ui| {
                self.render(ui);
            });
        }
    }
}
```

Gives me this result below. I would have expected enlarge the eframe height.

<img width="1348" height="1073" alt="AdaptToInner02" src="https://github.com/user-attachments/assets/efdcfbd2-9c85-4cb2-8f9f-b2939fd05c10" />

**Reply by liuzhch1** [reply]
hi, at that time I was using 0.30.0, which can be seen in the permalink project above
I haven't looked at this issue since then...

**Reply by PJaros** [reply]
I've changed now my `Cargo.toml` file to
```toml
eframe = "=0.30.0"
egui = "=0.30.0"
```
then I've run
```bash
cargo clean
cargo run
```
But the result is still the same.

I'm running on Wayland.
```bash
❯ echo $XDG_SESSION_TYPE
wayland
```

Edit:
Removed my question which OS you are using and the request for a screenshot.

I only now discovered the screenshot on your github project at https://github.com/liuzhch1/eframe-window-size-fit-content. So your code works on MacOS. And it works as expected.

Can @emilk give us a idea how to make this code work under Linux? Will this work under Windows?
I haven't tried to cross-compile with rust just yet.

**Reply by PJaros** [reply]
I found a solution. Rendering the window, fetching the size and setting the CenterPanel size need each a new round of update(...) cycles to make it work on my setup.

I'm cleaning up my code before posting it here.

**Comment by PJaros** [comment]
I've created a repo with the code: https://github.com/PJaros/eframe-fit-content

```bash
use eframe::egui;
use eframe::egui::{Vec2, Visuals, vec2};

fn main() -> eframe::Result {
    let app = AppGUI::new();
    eframe::run_native(
        "eframe fit content",
        eframe::NativeOptions::default(),
        Box::new(|_cc| Ok(Box::new(app))),
    )
}

enum Stage {
    PreRender(isize),
    FirstRender(Vec2),
    FirstResize(Vec2),
    Initialized(Vec2),
}

struct AppGUI {
    _render_stage: Stage,
}

impl AppGUI {
    fn new() -> Self {
        Self {
            _render_stage: Stage::PreRender(2_isize),
        }
    }
}

impl AppGUI {
    fn pre_render(&mut self, ctx: &eframe::egui::Context) {
        egui::Window::new("pre_render")
            .title_bar(false)
            .fixed_pos((0.0, 0.0))
            .show(ctx, |ui| {
                self.render(ui);
            });
    }
    fn render(&mut self, ui: &mut egui::Ui) {
        ui.add(
            egui::Label::new(egui::RichText::new("Hello, world!").heading())
                .wrap_mode(egui::TextWrapMode::Extend),
        );
        ui.label(format!(
            "self._used_size: {}",
            match self._render_stage {
                Stage::PreRender(_) => vec2(123.012_344, 123.012_344),
                Stage::FirstRender(size) => size,
                Stage::FirstResize(size) => size,
                Stage::Initialized(size) => size,
            }
        ));
        for n in 1..=10 {
            ui.label(format!("Line: {}", n));
        }
    }
}

impl eframe::App for AppGUI {
    fn update(&mut self, ctx: &eframe::egui::Context, _frame: &mut eframe::Frame) {
        ctx.set_visuals(Visuals::light());
        match self._render_stage {
            Stage::PreRender(mut pre_render_cycle) => {
                self.pre_render(ctx);
                pre_render_cycle -= 1;
                if pre_render_cycle > 0 {
                    self._render_stage = Stage::PreRender(pre_render_cycle)
                } else {
                    self._render_stage = Stage::FirstRender(ctx.used_size());
                }
            }
            Stage::FirstRender(size) => {
                egui::CentralPanel::default().show(ctx, |ui| {
                    self.render(ui);
                });
                self._render_stage = Stage::FirstResize(size)
            }
            Stage::FirstResize(size) => {
                ctx.send_viewport_cmd(egui::ViewportCommand::InnerSize(size));
                self._render_stage = Stage::Initialized(size)
            }
            Stage::Initialized(_size) => {
                egui::CentralPanel::default().show(ctx, |ui| {
                    self.render(ui);
                });
            }
        }
    }
}
```

The GUI looks like this:

<img width="240" height="249" alt="eframe-fit-content_scr01_50" src="https://github.com/user-attachments/assets/a19e1bf7-5e7f-4b34-8721-5d20007bf24d" />

Edit: Changed the code to reflect to current (less messy) state on the repo.
Edit2: Changed the code again to allow multiple pre-render cycles. This was needed in my other repo here: https://github.com/PJaros/menu_pdf
